### PR TITLE
baseApiUrl w/ path & normalizeTileURL functionality

### LINKED ISSUE #8458: baseApiUrl w/ path & normalizeTileURL functionality
**mapbox-gl-js version**: v0.54 / 1.0.0 / master

**browser**: chrome

I am experiencing malformed URLs generated from `normalizeTileURL` and `makeAPIURL`. This issue seems to only occur when the baseAPIURL contains a path (as opposed to just a host). I want to compare the behavior between mapbox.com & an Atlas hosted solution w/ a path of `https://localhost:8080/mbx`.

### .com
```
https://api.mapbox.com/v4/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7.json
```
is returning this tile format string:
```
https://a.tiles.mapbox.com/v4/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7/{z}/{x}/{y}.vector.pbf
```
It enters the `normalizeTileURL` function as:
```
mapbox://tiles/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7/10/184/401.vector.pbf
```
manipulated within the function to
```
/v4/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7/10/184/401.vector.pbf
```
and finally, from `makeAPIURL`:
```
https://api.mapbox.com/v4/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7/10/184/401.vector.pbf
```

### Atlas Hosted
```
http://localhost:8080/mbx/v4/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7.json
```
is returning me a semi-accurate tile format string (the protocol is wrong):
```
https://localhost:8080/mbx/v4/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7/{z}/{x}/{y}.vector.pbf
```
It enters the `normalizeTileURL` function this way, however inside this function the path is manipulated, and the `v4` prefix is needlessly appended:
```
/v4/mbx/v4/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7/10/184/401.vector.pbf
```
In the return statement this urlObject is passed to `makeAPIURL`, for a final url of
```
"http://localhost:8080/mbx/v4/mbx/v4/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7/10/184/401.vector.pbf"
```
You can see we got the API path prepended again.

## Seeking Guidance
I don't understand the rewriting of the url format provided by the `/v4` request. Additionally it looks like we are losing the subdomaining browser hack for increased parallelized requests?

### PR DESCRIPTION
**Pull request by sumarlidason**
- adds regex to trim tileURL prefix to /v4/...
- adds filter to makeAPIURL preventing duplication of param 'access_token'
- adds test for edge case

closes #8458

## Launch Checklist
 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page
 - [ ] tagged `@mapbox/studio` and/or `@mapbox/maps-design` if this PR includes style spec changes
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port

### CODE REVIEW

File: test/unit/util/mapbox.test.js [hunk 1]
```diff
@@ -357,6 +357,15 @@ test("mapbox", (t) => {
                 t.end();
             });
 
+            t.test('.normalizeTileURL ignores non-mapbox:// sources', (t) => {
```
Comment (ryanhamley) [inline test/unit/util/mapbox.test.js]: This test description looks like it was copied from the test above it

File: src/util/mapbox.js [hunk 2]
```diff
@@ -110,13 +110,15 @@ export class RequestManager {
 
         const urlObject = parseUrl(tileURL);
         const imageExtensionRe = /(\.(png|jpg)\d*)(?=$)/;
+        const tileURLAPIPrefixRe = /^.+\/v4\//;
```
Comment (ryanhamley) [inline src/util/mapbox.js]: Can you add a comment here that explains why this is necessary?

**Comment by sumarlidason** [comment]
@ryanhamley I explained this a bit in the issue; when the TileHost metadata from the api returns with a path e.g. `/mbx/`, the URI passed to normalizeTileURL is not formatted into a mapbox namespace `tilesURL: "mapbox://tiles/..."` as you see here (from an api w/o a path):
Screen Shot 2019-07-18 at 10 07 31 AM

With a path the URI looks like this,
Screen Shot 2019-07-18 at 10 09 29 AM

You can see that w/ the logic prior to this PR the final URL is malformed w/ duplicate `/mbx` and `/v4`. 

Another possible fix for this would be to adjust whatever parses this response's `resp.tiles[0]` into the mapbox namespace accounting for tile hosts w/ paths (`mapboxgl.baseApiUrl`):
image

I suspect that if the URI was properly coerced to the mapbox namespace this function would continue to function w/o modifications. I am not familiar enough with mapbox-gl-js to be aware of the ramifications -

File: src/util/mapbox.js (hunk 2 above)
Comment (sumarlidason) [reply src/util/mapbox.js]: https://github.com/mapbox/mapbox-gl-js/pull/8466#issuecomment-512837197

File: test/unit/util/mapbox.test.js (hunk 1 above)
Comment (sumarlidason) [inline test/unit/util/mapbox.test.js]:
```suggestion
            t.test('.normalizeTileURL accounts for tileURLs w/ paths', (t) => {
```

**Comment by sumarlidason** [comment]
@ryanhamley can you comment back and let me know what you think before me merge. 💯

**Comment by ryanhamley** [comment]
Yep looks good!
